### PR TITLE
Fully Updated TRPG tutorial (including 4.0 codeport) up to 'The Flood Fill Algorithm`

### PR DESCRIPTION
**Pull request by TheYellowArchitect**
Hello!
After spending many years with Unity, I moved to Godot. I have been stalking it since 3.x and being hyped with 4.0, when beta released, I finally installed it.

Doing a 3.x tutorial, is an opportunity to port it to 4.0, and I did this for the official godot 2D tutorial and the [3D tutorial (Squash the Creeps)](https://github.com/godotengine/godot-docs/pull/6243)

Next tutorial? Tactical Turn Based, and I bet I would be able to do this here too. Sadly, the progress was at a snail's pace, but there was progress nonetheless. I want to upload all my updates, instead of having them locally (because in a week I will have forgotten about them and delete them), and provide much feedback, as a new user (literally Q&A feedback lol)

Huge problem in the tutorial is the code. As in, everything is easy to follow, then suddenly there is a random wall of text of 50+ lines. Most of the functionality is unexplained. For example, [setget, or how tool functions really work](https://www.gdquest.com/tutorial/godot/2d/tactical-rpg-movement/lessons/02.the-unit/#coding-the-unit)

Basically, you should give small code snippets and explain them like the docs do, and keep doing this, but instead in this case,  there is an entire class dump, and the user spends ages. Because sure, the comments are good, but understanding *how* they all connect, does take time which should be mentioned by the tutorial.

Second huge problem is that you ask users to pull tilemaps and tilesets which are already made. Instead of making them in the tutorial, and the user learning how tilemaps/tilesets work, you cheat them of how to create those. I made an arrow unit path, and I had the map, but I did not understand anything on how tilemaps or how the grid *visually* works.

Those are huge problems above. Now let me go to the minor ones:

- **Lack of images makes some steps hard to follow.** 
For example, RAW color, how to even access that? Where do I click? Just show the image. This is mostly in the early sections of the tutorial. But it is present later too, see https://www.gdquest.com/tutorial/godot/2d/tactical-rpg-movement/lessons/04.pathfinding-and-path-drawing/ first comment. New users don't know how to create new scripts which are unrelated to nodes.
It's a minor problem because I uploaded images for these gaps, and this problem doesn't exist anymore (exception is the above example as I got frustrated with pathfinding and just went 'fuck it git push ggwp')

- **Visual Scaling**
e.g. the tilemap imported being tiny even with 0.25 scale player, so the visual instructions are off. I did my best to make them properly-sized.

- **3.x incompatible code**
Really takes time to find 4.0 alternatives, but I cannot blame you at all, its a 3.5 tutorial after all. However, it often took me much time to find the 4.0 equivalent (e.g. setget with custom setter), and at the final section I reached I straight up gave up because the Array type had issues casting to Vector2 or sth (_grid is null, so not even the compiler can aid me in what is wrong)

- **Use of Vector2 instead of Vector2i**
See code in https://www.gdquest.com/tutorial/godot/2d/tactical-rpg-movement/lessons/01.grid/
Vector2i is ideal for this tutorial, being simpler to the user as well. Not to mention you skip the .floor function in the above link/code, which means 1 less function.

- **grid.clamp() instead of using the native Variant.clamp()**
Variant.clamp() in 4.0 does what grid.clamp() does. Anyway, not a big problem, grid.clamp() could stay as it is and return clamp(grid_position, Vector2.ZERO, Vector2(size.x-1,size.y-1))

- **Unintuitive naming**
Bad naming with 'Sprite' baited me strongly (as there were no images to correct me either), so I renamed them to 'ShadowSprite' and 'CharacterSprite'
Some variables are also renamed as they are bad. `cell` to `hoveringCell` for Cursor.gd and `cell` to `occupyingCell` for Unit.gd

- **Unable to understand comment**
In https://www.gdquest.com/tutorial/godot/2d/tactical-rpg-movement/lessons/02.the-unit/ in "smooth movement logic"
```python
if not Engine.editor_hint:
		# We create the curve resource here because creating it in the editor prevents us from
		# moving the unit.
		curve = Curve2D.new()
```
What does the above comment mean?
I also added some useful comments here and there, where needed.

===========================

Ultimately, I upload what I did (alongside the reproduction project [trpg.tar.gz](https://github.com/GDQuest/website/files/9723760/trpg.tar.gz)), so you can update this tutorial easier to 4.0 and update many of its parts with images and clarifications. I also posted feedback so you can improve this tutorial and others (text tutorial having random walls of text is bad), because I have even made my own online level editor in Unity and am familiar with astar too, yet I didn't understand anything on tilemaps! It was just magic "use this mysterious native function bro" "import that .tres asset bro", which frustrated me. Same goes for `Path2D` logic which was unexplained, e.g. how the curve really works. Regardless, I hope my feedback is useful.

Anyway, I want to give more feedback on what a new user will seek to get from this kind of tutorial:
- Mouse Raycast onto any tilemap to get its tile
- Convert a tilemap into a grid (same grid logic you made pretty much). More accurately layer any tilemap with any grid ;)
- How to have a 2D tilemap in 3D space (via camera)

Also please say you will use tilemaps in the grid section because I had finished half of the tutorial, and I thought you would render lines for the grid, and then suddenly "import tilemap" happened which surprised me.

I guess I will go tinker by myself trial&error with tilemap class, maybe make chess in godot to learn tilemaps (there is not a single 4.0 tilemap grid tutorials in existence reeeeeeeeeeee)

### CODE REVIEW

**Comment by TheYellowArchitect** [comment]
Thank you for your swift reply.

>However, we made this series and a bunch of others for experienced Godot users and developers. Hence the long snippets. It can surely be improved, but breaking down long classes into short snippets and having a lot more detail is a trade-off: the series becomes more accessible, but it becomes way less interesting for experienced people who want the highlights and don't need small digestible steps.

In fact, having more code snippets makes it easier for experienced people to take the highlights, but anyway this is subjective and depends on the tutorial, and the code. I would still argue for more clarification, as that comment I highlighted shows I am not alone. And related to this is that most users finding these tutorials are new users. GDQuest is top link in tutorial resources, alongside most search results, hence saying "we care about experienced users" is so weird. This is a tutorial, not a demo. The goal isn't getting stuff done, but understanding how/why it works.

One could use the same logic of "veterans-only" to have less images since its "obvious" where to click for X menu. And imo veterans won't bother with an entire tutorial but just get a certain part. This was at least the case for me with Unity, after a certain point I never did another tutorial and could do anything as I was familiar with all engine components and API.

tl;dr of the above: The majority of your audience is newbies (especially with 4.0), so you can't use "we don't care about newbies" as an excuse for having less images and details lol

>Right now, we don´t want existing Godot 3 tutorials replaced with Godot 4 ports. The Godot 4 version would first have to be a different series. Also, it's an opportunity to change and upgrade the project rather than port it. We can do better than this.

I agree. It's a lot of tutorials to replace, and having a "3.x" and "4.x" branch support is hard. I suggest trying to remake the tutorials, but starting from scratch. This one for example could have a section or two on tilemaps and how it works on code. Same for `path`. With different assets, it would be truly a "new" tutorial just by copying the old lol

>Note that we've been at this for a long time, and we always work with our target audience and get feedback. So if you see something that doesn't work at all for you, it's likely that it wasn't created for you but for a different category of users who have different tastes and needs.

This applies to everything, may it be videogames, books, tutorials, opinions lol
I understand I am not the target audience, after all I did see it was made for 3.x
I did give it a try however, thinking I would at least reach the end, and it's not your tutorial to blame, but me thinking 4.0 to 3.x is mere renames. I had to upload my efforts regardless, because even if you do deny this pull request (its incomplete, so its logical), the feedback is hopefully useful for future tutorials.

Off-topic: Is there any tutorial on the internet on the topic of coding tilemaps in 4.0? Basically I want to have a grid (2D array), and generate tilemaps for it via code, and also be able to raycast from any (3D) camera to get a tile index. If I understand that, I am certain I can figure out everything else

**Comment by TheYellowArchitect** [comment]
I misunderstood the 'series' with the entire GDQuest part. I apologize for that, thinking it was a deflection of all criticism lol

>lots of seasoned users and professionals who need learning material made for them
If someone is seasoned or professional, I doubt he would pay money to learn something he can trial&error and get working in a week. I do not write this to undermine or insult, but certainly some of that 50K$ was by users who want Godot to succeed and improve, regardless of skill.

Anyway, the tutorial helped me with the overall idea of how to layout a TRPG, i will go tinker with `Path2D` until I figure out how to code my own tilemap, I wish you good future tutorials :+1:
